### PR TITLE
Refactor the parser in a single NetscapeBookmarkParser class

### PR DESCRIPTION
**Pull request by virtualtam**
Modifications:
- update Composer autoload
- encapsulate all methods in a single class
- update tests
- update README and example

### CODE REVIEW

**Comment by virtualtam** [comment]
Ping @ArthurHoaro & @nodiscc , as this parser should be mature enough to be used within Shaarli ;-)

File: NetscapeBookmarkParser.php [hunk 1]
```diff
@@ -0,0 +1,222 @@
+<?php
+
+/**
+ * Generic Netscape bookmark parser
+ */
+class NetscapeBookmarkParser
+{
+    protected $defaultTag = '';
+    protected $defaultPub = false;
+    protected $items = array();
```
Comment (ArthurHoaro) [inline NetscapeBookmarkParser.php]: Default values are useless for attributes set in the constructor, which is a best practice.

File: NetscapeBookmarkParser.php [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+     */
+    public static function sanitizeString($bookmarkString)
+    {
+        $sanitized = $bookmarkString;
+        $sanitized = str_replace(array("\r", "\t"), array('',' '), $sanitized);
+
+        $sanitized = preg_replace_callback(
+            '@<DD>(.*?)(<A|<\/|<DL|<DT|<P)@mis',
+            function($match) {
+                return '<DD>'.str_replace("\n", '<br>', trim($match[1])).'</';
```
Comment (ArthurHoaro) [inline NetscapeBookmarkParser.php:208]: Aren't you losing `<A|<DL|<DT|<P` here?

File: NetscapeBookmarkParser.php [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                    $this->items[$i]['tags'] = $currentTag;
+                }
+
+                if (preg_match('/href="(.*?)"/i', $line, $m3)) {
+                    $this->items[$i]['uri'] = $m3[1];
+                } else {
+                    $this->items[$i]['uri'] = '';
+                }
+
+                if (preg_match('/<a(.*?)>(.*?)<\/a>/i', $line, $m4)) {
```
Comment (ArthurHoaro) [inline NetscapeBookmarkParser.php]: Unnecessary first capture.

File: NetscapeBookmarkParser.php [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                    $this->items[$i]['title'] = $m4[2];
+                } else {
+                    $this->items[$i]['title'] = 'untitled';
+                }
+
+                if (preg_match('/note="(.*?)"<\/a>/i', $line, $m5)) {
```
Comment (ArthurHoaro) [inline NetscapeBookmarkParser.php:108]: `note=` doesn't exists in your example files. Do you have other examples?

File: composer.json [hunk 5]
```diff
@@ -1,10 +1,10 @@
 {
     "name": "kafene/netscape-bookmark-parser",
-    "description": "Parse Netscape format bookmark files",
+    "description": "Generic Netscape bookmark parser",
     "keywords": ["parse", "bookmark", "netscape"],
     "homepage": "https://github.com/kafene/BookmarkParser",
     "license": "MIT",
-    "autoload": {"files": ["parse_netscape_bookmarks.php"]},
+    "autoload": {"files": ["NetscapeBookmarkParser.php"]},
     "require": {},
```
Comment (ArthurHoaro) [inline composer.json:8]: `"php": ">=5.3"`

File: NetscapeBookmarkParser.php [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        $i = 0;
+        $next = false;
+        $currentTag = $this->defaultTag;
+
+        $lines = explode("\n", $this->sanitizeString($bookmarkString));
+
+        foreach ($lines as $line_no => $line) {
+            if (preg_match('/^<h\d(.*?)>(.*?)<\/h\d>/i', $line, $m1)) {
+                // a tag is matched: set its value as current
+                $currentTag = $m1[2];
```
Comment (ArthurHoaro) [inline NetscapeBookmarkParser.php:83]: Looking at https://github.com/shaarli/netscape-bookmark-parser/blob/shaarli-devel/tests/input/shaarli.htm#L7, I'm not sure that's a good idea.

File: NetscapeBookmarkParser.php [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                     $this->items[$i]['note'] = str_replace('<br>', "\n", $m6[1]);
+                } else {
+                    $this->items[$i]['note'] = '';
+                }
+
+                if (preg_match('/(tags?|labels?|folders?)="(.*?)"/i', $line, $m7)) {
+                     $this->items[$i]['tags'] = strtr($m7[2], ',', ' ');
+                } else {
+                    $this->items[$i]['tags'] = $currentTag;
+                }
```
Comment (ArthurHoaro) [inline NetscapeBookmarkParser.php:120]: If I ask for a default tag, it might be a good idea to append it either it finds existing tags or not?

**Comment by ArthurHoaro** [comment]
Great! :+1: 
I've asked a few question about how it handles tags, but almost ready.

Also we'll have to be careful about empty URL and default timestamp with Shaarli integration.

File: NetscapeBookmarkParser.php (hunk 4 above)
Comment (virtualtam) [inline NetscapeBookmarkParser.php:108]: Sadly, no; it's quite hard to even find bookmark dumps for "active" services such as Delicious (that is, without registering to them).

File: NetscapeBookmarkParser.php (hunk 7 above)
Comment (virtualtam) [inline NetscapeBookmarkParser.php:120]: That would be relevant, yet barely used IMHO as I'd rather keep an empty tag list... empty ^^

File: NetscapeBookmarkParser.php (hunk 2 above)
Comment (virtualtam) [inline NetscapeBookmarkParser.php:208]: There's a bit of magic involved here, as well as in the main loop, with a mix of single- and multi-line regex flags. 

What this method (supposedly) does is converting all data related to a given link so it is represented in a single line, with `<br>` for line breaks, hence the main parsing loop iterating on the resulting "lines".

It feels like the sanitizing/trimming ops could be further simplified to gain in readability.

File: NetscapeBookmarkParser.php (hunk 6 above)
Comment (virtualtam) [inline NetscapeBookmarkParser.php:83]: Good catch! The parsed data might need an additional section to handle global metadata (or simply ditch it)

File: composer.json (hunk 5 above)
Comment (virtualtam) [inline composer.json:8]: #15 raised to improve packaging

File: NetscapeBookmarkParser.php (hunk 1 above)
Comment (virtualtam) [inline NetscapeBookmarkParser.php]: done ;-)

**Comment by virtualtam** [comment]
Thanks for the review!

Regarding Shaarli-specific formatting:
- there are some notes in the test data (the "URL" only is the permalink part)
- I think there is a hack on the timestamps in Shaarli's `index.php`, that allows importing several links with the same timestamp

I think it'd also be worth simplifying the parser's cleanup steps, all those regexes feel a bit superfluous (=> separate PR)

**Comment by ArthurHoaro** [comment]
Permalinks are fine, we have to support it, to move a Shaarli to another for example. I was talking about [this one](https://github.com/shaarli/netscape-bookmark-parser/pull/14/files#diff-4a59dbd62a2dae2500f43e35fb7d51ceR99).

**Comment by virtualtam** [comment]
> I was talking about this one.

Hmm, I wouldn't worry too much about "linkless links", the only way one would end with this is by manually editing a bookmark dump and deleting the `href` attribute, which is kinda silly...

**Comment by virtualtam** [comment]
To sum up:
- #15 to improve packaging metadata
- #16 to add test cases & improve robustness a little bit
- #17 to simplify data cleanup (and maybe remove dead/useless code)

**Pull request merged**